ExtendedQueryCommandBaseCodec not close prepare statement while receive an error packet

### Questions

while execute prepare sql  encounter an error, more info errorCode=1062  sqlState=23000， client will received an error packet. Then, we miss to close prepare statement, see the function handleErrorPacketPayload.

### Version

4.2.4

I was able to reproduce with:

```java

  @Test
  public void testPreparedStatementCleaned(TestContext ctx) {
    Assume.assumeFalse(MySQLTestBase.rule.isUsingMySQL5_6() || MySQLTestBase.rule.isUsingMariaDB());
    MySQLConnectOptions connectOptions = new MySQLConnectOptions(options)
      .setUser("root")
      .setPassword("password")
      .setCachePreparedStatements(false);
    Async async = ctx.async();
    MySQLConnection.connect(vertx, connectOptions, ctx.asyncAssertSuccess(conn -> {
      conn.query("SELECT * FROM performance_schema.prepared_statements_instances").execute(ctx.asyncAssertSuccess(res1 -> {
        ctx.assertEquals(0, res1.size());
        conn.preparedQuery("INSERT INTO duplicate_test VALUES (?)").execute(Tuple.of(1), ctx.asyncAssertFailure(failure -> {
          if (!(failure instanceof MySQLException)) {
            ctx.fail(failure);
            return;
          }
          MySQLException e = (MySQLException) failure;
          ctx.assertEquals(1062, e.getErrorCode());
          ctx.assertEquals("23000", e.getSqlState());
          conn.query("SELECT * FROM performance_schema.prepared_statements_instances").execute(ctx.asyncAssertSuccess(res3 -> {
            ctx.assertEquals(0, res3.size()); // ! Fails here
            conn.close(ctx.asyncAssertSuccess(v -> {
              async.complete();
            }));
          }));
        }));
      }));
    }));
  }
```

Which fails with

```
java.lang.AssertionError: Not equals : 0 != 1
```

--- a/vertx-mysql-client/src/main/java/io/vertx/mysqlclient/impl/codec/ExtendedQueryCommandBaseCodec.java
+++ b/vertx-mysql-client/src/main/java/io/vertx/mysqlclient/impl/codec/ExtendedQueryCommandBaseCodec.java
@@ -35,6 +35,7 @@ abstract class ExtendedQueryCommandBaseCodec<R, C extends ExtendedQueryCommand<R
       OkPacket okPacket = decodeOkPacketPayload(payload);
       handleSingleResultsetDecodingCompleted(okPacket.serverStatusFlags(), okPacket.affectedRows(), okPacket.lastInsertId());
     } else if (firstByte == ERROR_PACKET_HEADER) {
+      closePreparedStatement();
       handleErrorPacketPayload(payload);
     } else {
       handleResultsetColumnCountPacketBody(payload);
@@ -43,12 +44,15 @@ abstract class ExtendedQueryCommandBaseCodec<R, C extends ExtendedQueryCommand<R
 
   @Override
   protected void handleAllResultsetDecodingCompleted() {
-    // Close prepare statement
+    closePreparedStatement();
+    super.handleAllResultsetDecodingCompleted();
+  }
+
+  private void closePreparedStatement() {
     MySQLPreparedStatement ps = (MySQLPreparedStatement) this.cmd.ps;
     if (ps.closeAfterUsage) {
       sendCloseStatementCommand(ps);
     }
-    super.handleAllResultsetDecodingCompleted();
   }
 
   private void sendCloseStatementCommand(MySQLPreparedStatement statement) {

--- a/vertx-mysql-client/src/test/java/io/vertx/mysqlclient/MySQLPreparedStatementTest.java
+++ b/vertx-mysql-client/src/test/java/io/vertx/mysqlclient/MySQLPreparedStatementTest.java
@@ -173,4 +173,34 @@ public class MySQLPreparedStatementTest extends MySQLTestBase {
       }));
     }));
   }
+
+  @Test
+  public void testPreparedStatementCleaned(TestContext ctx) {
+    Assume.assumeFalse(MySQLTestBase.rule.isUsingMySQL5_6() || MySQLTestBase.rule.isUsingMariaDB());
+    MySQLConnectOptions connectOptions = new MySQLConnectOptions(options)
+      .setUser("root")
+      .setPassword("password")
+      .setCachePreparedStatements(false);
+    Async async = ctx.async();
+    MySQLConnection.connect(vertx, connectOptions, ctx.asyncAssertSuccess(conn -> {
+      conn.query("SELECT * FROM performance_schema.prepared_statements_instances").execute(ctx.asyncAssertSuccess(res1 -> {
+        ctx.assertEquals(0, res1.size());
+        conn.preparedQuery("INSERT INTO duplicate_test VALUES (?)").execute(Tuple.of(1), ctx.asyncAssertFailure(failure -> {
+          if (!(failure instanceof MySQLException)) {
+            ctx.fail(failure);
+            return;
+          }
+          MySQLException e = (MySQLException) failure;
+          ctx.assertEquals(1062, e.getErrorCode());
+          ctx.assertEquals("23000", e.getSqlState());
+          conn.query("SELECT * FROM performance_schema.prepared_statements_instances").execute(ctx.asyncAssertSuccess(res2 -> {
+            ctx.assertEquals(0, res2.size());
+            conn.close(ctx.asyncAssertSuccess(v -> {
+              async.complete();
+            }));
+          }));
+        }));
+      }));
+    }));
+  }
 }

--- a/vertx-mysql-client/src/test/resources/init.sql
+++ b/vertx-mysql-client/src/test/resources/init.sql
@@ -249,3 +249,13 @@ INSERT INTO collector_test
 VALUES (2, 32767, 2147483647, 9223372036854775807, 123.456, 1.234567, 'hello,world');
 
 -- TCK usage --
+
+DROP TABLE IF EXISTS duplicate_test;
+CREATE TABLE duplicate_test
+(
+  primary_key INT NOT NULL,
+  CONSTRAINT duplicate_test_pk
+    PRIMARY KEY (primary_key)
+);
+INSERT INTO duplicate_test
+VALUES (1);

